test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/TestFhirServer.cs

```diff
@@ -52,26 +69,125 @@ public TestFhirClient GetTestFhirClient(ResourceFormat format, TestApplication c
 
             if (!reusable)
             {
-                return CreateFhirClient(format, clientApplication, user);
+                return CreateFhirClient(format, clientApplication, user, authenticationHandler);
             }
 
             return _cache.GetOrAdd(
                     (format, clientApplication, user),
                     (tuple, fhirServer) =>
-                        new Lazy<TestFhirClient>(() => CreateFhirClient(tuple.format, tuple.clientApplication, tuple.user)),
+                        new Lazy<TestFhirClient>(() => CreateFhirClient(tuple.format, tuple.clientApplication, tuple.user, authenticationHandler)),
                     this)
                 .Value;
         }
 
-        private TestFhirClient CreateFhirClient(ResourceFormat format, TestApplication clientApplication, TestUser user)
+        private TestFhirClient CreateFhirClient(ResourceFormat format, TestApplication clientApplication, TestUser user, DelegatingHandler authenticationHandler = null)
         {
-            var httpClient = new HttpClient(new SessionMessageHandler(CreateMessageHandler(), _asyncLocalSessionTokenContainer)) { BaseAddress = BaseAddress };
+            ConfigureSecurityOptions().GetAwaiter().GetResult();
+
+            var sessionMessageHandler = new SessionMessageHandler(CreateMessageHandler(), _asyncLocalSessionTokenContainer);
+
+            if (authenticationHandler != null)
+            {
+                authenticationHandler.InnerHandler = CreateMessageHandler();
+                sessionMessageHandler.InnerHandler = authenticationHandler;
+            }
+            else if (SecurityEnabled && !clientApplication.Equals(TestApplications.InvalidClient))
+            {
+                string authKey = GenerateKey();
+                if (_authenticationHandlers.ContainsKey(authKey))
+                {
+                    sessionMessageHandler.InnerHandler = _authenticationHandlers[authKey];
+                }
+                else
+                {
+                    AuthenticationHttpMessageHandler authHandler;
+
+                    var authHttpClient = new HttpClient(new SessionMessageHandler(CreateMessageHandler(), _asyncLocalSessionTokenContainer)) { BaseAddress = BaseAddress };
+                    if (user == null)
+                    {
+                        string scope = clientApplication.Equals(TestApplications.WrongAudienceClient) ? clientApplication.ClientId : AuthenticationSettings.Scope;
```

Comment:
I think there are a little bit too much logic (or test specific logic?) in this method. Would it be more readable if we separate these out?

For example, there is CreateFhirClient(ResourceFormat format, DelegatingHandler authenticationHandler) that creates a FhirClient with whatever the handler that's passed in. There could be another method that reads the CapabilityStatement and use the clientApplication/user to create an appropriate handler and call the previous method. There could be another method that creates handler using an invalid application id and call the previous method. Just to separate the logic out a litte?